### PR TITLE
pbr: initial commit

### PR DESCRIPTION
**Pull request by stangri**
Maintainer: me
Compile tested: x86_64, Sophos SG-135, OpenWrt 22.03.2
Run tested: x86_64, Sophos SG-135, OpenWrt 22.03.2, test basic functionality with dnsmasq.nftset and simple policies

Description
* The makefile produces the nft and iptables capable `pbr` package and the `pbr-iptables` package for legacy setups
* This replaces `vpnbypass` and `vpn-policy-routing` packages
* I'm soliciting feedback on this package and my intention is to update the version to 1.0.0 before this is merged, but I need the feedback on this and luci-app-pbr before then.

Signed-off-by: Stan Grishin <stangri@melmac.ca>

### CODE REVIEW

File: net/pbr/Makefile [hunk 1]
```diff
@@ -0,0 +1,198 @@
+# Copyright 2017-2022 Stan Grishin (stangri@melmac.ca)
+# This is free software, licensed under the GNU General Public License v3.
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=pbr
+PKG_VERSION:=0.9.9
+PKG_RELEASE:=24
```
Comment (BKPepe) [inline net/pbr/Makefile:8]:
```suggestion
PKG_RELEASE:=1
```

File: net/pbr/Makefile [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+define Package/pbr/default
+	SECTION:=net
+	CATEGORY:=Network
+	SUBMENU:=VPN
+	PROVIDES:=pbr
+	TITLE:=Policy Based Routing Service
+	URL:=https://docs.openwrt.melmac.net/pbr/
+	DEPENDS:=+ip-full +jshn +jsonfilter +resolveip
+	CONFLICTS:=vpnbypass vpn-policy-routing
```
Comment (BKPepe) [inline net/pbr/Makefile:22]: Shouldn't pbr provide these packages?

File: net/pbr/Makefile [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	PKGARCH:=all
+endef
+
+define Package/pbr
+$(call Package/pbr/default)
```
Comment (BKPepe) [inline net/pbr/Makefile:27]: Why this package provides itself based from ``define Package/pbr/default``?

File: net/pbr/Makefile [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+PKG_NAME:=pbr
+PKG_VERSION:=0.9.9
+PKG_RELEASE:=24
+PKG_LICENSE:=GPL-3.0-or-later
+PKG_MAINTAINER:=Stan Grishin <stangri@melmac.ca>
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/pbr/default
```
Comment (BKPepe) [inline net/pbr/Makefile:14]: This section should has two spaces as indentation.

File: net/pbr/Makefile [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+This version supports OpenWrt with fw3/ipset/iptables.
+endef
+
+define Package/pbr-netifd/description
+This service enables policy-based routing for WAN interfaces and various VPN tunnels.
+This version supports OpenWrt with both fw3/ipset/iptables and fw4/nft.
+This version uses OpenWrt native netifd/tables to set up interfaces. This is WIP.
+endef
+
+define Package/pbr/default/conffiles
```
Comment (BKPepe) [inline net/pbr/Makefile:59]:
```suggestion
define Package/pbr/conffiles
```

and then you need to modify the conffiles below.

File: net/pbr/Makefile [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+Package/pbr-netifd/conffiles = $(Package/pbr/default/conffiles)
+
+define Build/Configure
+endef
+
+define Build/Compile
+endef
+
+define Package/pbr/default/install
+	$(INSTALL_DIR) $(1)/etc/config
```
Comment (BKPepe) [inline net/pbr/Makefile:74]: What rather to create directory first and then what files should be in that directory and etc? Right now, it is not very clear unless going through into it deeply what goes where.

File: net/pbr/files/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# README
```
Comment (BKPepe) [inline net/pbr/files/README.md:1]: Hmm, readme should be here or it should be removed.

**Review by BKPepe — CHANGES_REQUESTED** [review]

File: net/pbr/Makefile (hunk 1 above)
Comment (stangri) [reply net/pbr/Makefile:8]: > * I'm soliciting feedback on this package and my intention is to update the version to 1.0.0 before this is merged, but I need the feedback on this and luci-app-pbr before then.

**Comment by stangri** [comment]
@BKPepe thank you for your review, I'll address your suggestions within a few days!

File: net/pbr/Makefile (hunk 2 above)
Comment (stangri) [reply net/pbr/Makefile:22]: Will be fixed in the next push.

File: net/pbr/Makefile (hunk 3 above)
Comment (stangri) [reply net/pbr/Makefile:27]: What's a better way to organize two packages: `pbr` and `pbr-iptables` from the same Makefile?

File: net/pbr/Makefile (hunk 4 above)
Comment (stangri) [reply net/pbr/Makefile:14]: Thanks, will be fixed in the next push. It's just the package definition section which needs two spaces vs tabs for indentation, correct?

File: net/pbr/Makefile (hunk 6 above)
Comment (stangri) [reply net/pbr/Makefile:74]: Sorry, I'm not sure I'm getting your point, could you rephrase it?

File: net/pbr/files/README.md (hunk 7 above)
Comment (stangri) [reply net/pbr/files/README.md:1]: I strongly prefer to host the README and the screenshots it includes together, which is a no-go for this repo. Currently the README (and the screenshots) are hosted in another github repo and published thru github pages. 

Some people may discover/look for the README within the package directory that's why it's here.

File: net/pbr/Makefile (hunk 5 above)
Comment (stangri) [reply net/pbr/Makefile:59]: If I understood you correctly, this will be fixed in the next push.

File: net/pbr/Makefile (hunk 6 above)
Comment (oldium) [reply net/pbr/Makefile:74]: I think the idea was to group `INSTALL_DIR` with respective `INSTALL_DATA`/`BIN` so that it is clear which files go to which directories. Meaning to have something like this (like it is in `package/libs/libselinux/Makefile` for example):

```bash
$(INSTALL_DIR) $(1)/etc/init.d
$(INSTALL_BIN) ./files/etc/init.d/pbr.init $(1)/etc/init.d/pbr
$(SED) "s|^\(readonly PKG_VERSION\).*|\1='$(PKG_VERSION)-$(PKG_RELEASE)'|" $(1)/etc/init.d/pbr
$(INSTALL_DIR) $(1)/etc/hotplug.d/iface
$(INSTALL_DATA) ./files/etc/hotplug.d/iface/70-pbr $(1)/etc/hotplug.d/iface/70-pbr
…
```

This way it will be visible which directories are created empty for example.

Comment (BKPepe) [reply net/pbr/Makefile:74]: Hey @oldium, thanks! This is exactly what I meant.

Comment (stangri) [reply net/pbr/Makefile:74]: Thanks, I've reorganized that in 0.9.9-37 and will include this in the next push.
